### ISSUE TITLE
Graph Image not showing in Slack

### ISSUE THREAD

**Issue by MoonDogg** [jamesboswell/thold-slack#1]
The graph image is not showing in slack. I set the image URL and the image path that is mapped to the same URL location in Apache. I noticed the images are not being copied to the file location either. I am wondering if this has anything to do with the way the e-mails are formatted. (ex. inline images instead of attached.)

**Comment by jamesboswell** [comment]
Hi @MoonDogg,  Slack accesses the images via the URL path defined [here in line 42](https://github.com/jamesboswell/thold-slack/blob/master/thold-slack.py#L42)

It doesn't have to be the same server as your Cacti instance, but it needs to be a URL that Slack Image proxy can reach (https://api.slack.com/robots)

You then need to map the URL to the disk location defined in [line 58](https://github.com/jamesboswell/thold-slack/blob/master/thold-slack.py#L58)

Something like this 
```apache
Alias /img /data/tmp/img  # this maps /img URL to disk location
<Location /img>
    Allow from All
    Satisfy any
</Location>
```

**Comment by jamesboswell** [comment]
> I noticed the images are not being copied to the file location either. I am wondering if this has anything to do with the way the e-mails are formatted. (ex. inline images instead of attached.)

Check your file permissions on the directory defined in [line 58](https://github.com/jamesboswell/thold-slack/blob/master/thold-slack.py#L58)

**Comment by MoonDogg** [comment]
So I check the permission and I change them to 1777 as you suggested. They are currently owned by www-data but that should not matter with it set to 1777. The images are still not showing up in that directory. This is what my config looks like

Line 51 
`slack_image_url     =   "http://cacti.qhi.local/tmp/img/"
` 
This is currently only accessible from my internal network. But I will worry about that once I get the images showing up in the directory.  

Line 58 
`image_path           = '/var/www/html/tmp/img/'
`

**Comment by jamesboswell** [comment]
@MoonDogg can you try `image_path = '/tmp'` to just see if the image file writes at all?

`smrsh` is enabled in your Sendmail config, correct?  

```bash
# grep smrsh /etc/mail/sendmail.mc
FEATURE(`smrsh', `/usr/sbin/smrsh')dnl
```

**Comment by jamesboswell** [comment]
@MoonDogg if you can save an email,  you can manually test by 
```shell
$ cat email.eml | /etc/smrsh/thold-slack.py
```
and see what's going on

**Comment by MoonDogg** [comment]
I will do that... I'll let you know what I find out.

**Comment by MoonDogg** [comment]
This is what I got... 

> netmon@qhiserver16:/etc/mail/smrsh$ cat email.eml | ./thold-slack.py
> 
> POST to Slack.com successful
> 
> netmon@qhiserver16:/etc/mail/smrsh$ cd /var/www/html/tmp/img
> netmon@qhiserver16:/var/www/html/tmp/img$ ls
> netmon@qhiserver16:/var/www/html/tmp/img$ ls
> 

So it runs the script but no images...  

P.S. 
I want to say thanks for taking the time to look at this. I really do appreciate it. This is not a major problem for me I am super happy just to have the alerts show up in Slack. The graph is just an added bonus.

**Comment by MoonDogg** [comment]
I just saw the msg above the one I was originally responding to... and apparently smrsh is not enabled, or it is not in the config file you noted. I am going to investigate this further. Thanks.

**Comment by jamesboswell** [comment]
@MoonDogg were you able to resolve this issue?

**Comment by MoonDogg** [comment]
@jamesboswell no.. I never did get it working... and I don't remember where I left off at.. I don't think I ever got smrsh configured in sendmail. I have been real busy at work and this kinda got put on the back burner. 

I just checked my config and this is what is in there 
FEATURE(`smrsh')dnl

which I am assuming is incorrect because it has no path.

**Comment by jamesboswell** [comment]
@MoonDogg  in my `/etc/mail/sendmail.mc` file I have this line

```bash
FEATURE(`smrsh', `/usr/sbin/smrsh')dnl
```

This is on a RedHat/CentOS 6.x system.   After updating you have to run  
```bash
m4 /etc/mail/sendmail.mc > /etc/mail/sendmail.cf
```

Keep me posted.  FYI,  the thold-slack script has been updated to work better with multipart messages.   See  https://github.com/jamesboswell/thold-slack/releases

**Comment by MoonDogg** [comment]
Hmm I have ubuntu 16.04.02
```
whereis smrsh

smrsh: /usr/share/man/man8/smrsh.8.gz
```

Any idea what that is?

**Comment by jamesboswell** [comment]
@MoonDogg  should be part of `sendmail-bin` package  see:  https://packages.ubuntu.com/yakkety/amd64/sendmail-bin/filelist

**Comment by MoonDogg** [comment]
think I should edit my config to read 

`FEATURE(`smrsh', `/usr/share/man/man8/smrsh.8.gz')dnl`

**Comment by jamesboswell** [comment]
@MoonDogg   no,  should be  `/usr/lib/sm.bin/smrsh`  i believe,  see if that file exists

**Comment by MoonDogg** [comment]
Yeap sure does... ok I change it, see what happens..

**Comment by MoonDogg** [comment]
Hmm... I updated the script to your latest. change my sendmail config... then I change a Thold setting to make it trigger....  I got the e-mail but no slack alert....  I am heading home for the day... so I will look into this further tomorrow. Thanks for all your help  I will keep you updated on what I find.

**Comment by MoonDogg** [comment]
This is what the script does now...  any ideas?
```
/etc/mail/smrsh$ cat email.eml | ./thold-slack.py
Traceback (most recent call last):
  File "./thold-slack.py", line 241, in <module>
    main()
  File "./thold-slack.py", line 93, in main
    imgfile = get_image(mail)
  File "./thold-slack.py", line 216, in get_image
    image_part = image_parts[0]
IndexError: list index out of range
```


I put the original script back in and it works but still no graph in slack
```

:/etc/mail/smrsh$ cat email.eml | ./thold-slack.py
POST to Slack.com successful
```

**Comment by jamesboswell** [comment]
@MoonDogg I understand the error above and tracking it in #3 .   Please try this version of the script  https://github.com/jamesboswell/thold-slack/blob/issue3/thold-slack.py

I added some DEBUG lines,  I think your email message doesn't have any images?

**Comment by MoonDogg** [comment]
Ok so I updated the script with the one in your last comment and this is what happened.

```
/etc/mail/smrsh$ cat email.eml | ./thold-slack.py
DEBUG: No JPG subpart found!!
POST to Slack.com successful
```

The e-mail does have an image in it but its in PNG format. Is that an issue?

**Comment by jamesboswell** [comment]
@MoonDogg  if you go into "Plugin Management" what version of THOLD do you have installed?   `<your-server-address>/cacti/plugins.php`    I'm gonna guess it's 0.4, am I correct?

You an do a search and replace,  and replace `jpg` with `png`,  and try that.   I'd need to add some logic to make the script work with both JPG and PNG

In THOLD 0.4.6 it looks like it was changed to attach graphs as `image/jpg` instead of PNG, due to Lotus Notes not being able to render PNG
https://github.com/Cacti/plugin_thold/commit/dd6ce51e5c3b3174f658ebbc8bca0ae7bf665641?diff=split

**Comment by MoonDogg** [comment]
Thold Version is 1.0.2

I will try and edit the script and see what happens.

**Comment by MoonDogg** [comment]
Ok here is the outcome of the change. 

```
cat email.eml | ./thold-slack.py
DEBUG: png image graph_759.png subpart found
POST to Slack.com successful
```
But it did not show in Slack and I think I know why... a part of the script I have been ignoring with all the changes...  the image_path  let me look at older version to see what I had set before.. I'll let you know what happens..

**Comment by MoonDogg** [comment]
I think we are almost there... so I ran the script.. it ran correctly... the image file shows up in the folder I set.. and its accessible from the url I set in the script  (slack_image_url) the only thing I am thinking that may cause it not to work is, that url is not accessible from outside our network.  Is that an issue for slack?

**Comment by jamesboswell** [comment]
> the only thing I am thinking that may cause it not to work is, that url is not accessible from outside our network. Is that an issue for slack?

Yes that's an issue.  The URL you specify must be open to the Slack Imgproxy robot (https://api.slack.com/robots).   That's just how Slack works.

My installation `image_path` is a NFS dir mapped to a public URL on another server, 
`slack_image_url = <server.domain.com>/uncommon/url/img/`
that allows the Slack Imgproxy.   Web server configuration disallows any directory browsing,  and the image files are named with a random UUID, so it's very hard to find an image.

On top of that,  a cron job regularly deletes images (I don't need really old images)

In summary
1)  publicly reachable 'hidden' custom URL
2)  UUID random file name
3)  image files get deleted after X period

For me that's acceptable security,  but you should follow your own security policies.

The script could be modified to output the image file to something like an Amazon S3 bucket or Dropbox, etc. that Slack could reach.   Just not something I've needed.

**Comment by MoonDogg** [comment]
I figured that might be the case. Our website is externally hosted, and everything (except exchange) in our internal network is inaccessible from the outside world. I can setup a (server.domain.com) to point to our IP and route it to the cacti server. I'll think about how I want to implement this and test it out once I determine a course of action.

But from this point it looks like everything is working with the script. I appreciate all your help and I am willing to test any new changes you come up with.

**Comment by MoonDogg** [comment]
Update... I have configured everything in our hosted DNS and our routers, and now everything is works as it should. The graph's are showing up in Slack!! Thank you again for all your help!

**Comment by jamesboswell** [comment]
@MoonDogg  that's good to hear!  Closing this issue. Cheers.
